### PR TITLE
Enhance Quest Completion with Video Integration and Practice

### PR DESCRIPTION
**Pull request by ziyadbagh**
### PR Description

**Enhancements to Quest Completion with Camera Interaction**

This PR introduces significant improvements to the existing quest system, enriching the user's experience and functionality. Below are the key updates included:

- **Video Resource Integration:** @SaFouNaldo's recorded videos have been added to the database, providing users with visual aids tied directly to their quests. This ensures each quest has a clear and engaging reference point for users.

- **Fingerspelling Dialog Enhancement:** A refined dialog has been implemented to enable users to complete their daily quest interactively. Users can now use their camera to fingerspell the word of the day, providing a hands-on, immersive experience. The new design simplifies the process and aligns it more closely with user expectations (see the screenshot below for details).

- **Dynamic Video Fetching:** Introduced a `fetchVideo` function to dynamically retrieve the appropriate video from the database based on the quest. 

- **Testing Improvements:** Added dedicated tests for the new UI composables and the `fetchVideo` function to ensure the stability of the new and updated components. This helps maintain code quality and prevent regressions as the feature evolves.

**Next Steps:**
1. Expand the quest database by adding more words and their corresponding videos to provide a richer set of challenges for users.
2. Integrate user progress tracking by linking quest completion with the stats repository, enabling real-time updates and ensuring user achievements are recorded accurately.

These updates are designed to enhance the interactivity, scalability, and reliability of the quest system, paving the way for future extensions and improvements. Feedback and suggestions are highly encouraged!

imageimage

**Note:** @NBCCF025  removed the padding around the camera in a PR that has not been merged yet.

### CODE REVIEW

File: app/src/main/res/raw/hello.mp4
Comment (nada-ezziti) [inline app/src/main/res/raw/hello.mp4:1]: The letter H is missing in this video.

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
I tested your branch before diving into the code and noticed a few things that could be improved. 

First, I think it would be better to remove the background sound in the videos to make them more professional. 

Additionally, in the video for the word "Hello," the letter H is missing. I'll keep reviewing to see if I spot other missing letters. However, I suggest that before finalizing the app, we divide the list of words among the team to ensure all the videos are accurate and error-free.

I've submitted a change request for this, but I see you've already spent a lot of time on this feature. Let me know if you'd prefer to address these video-related improvements in a separate task.

File: app/src/main/java/com/github/se/signify/model/home/quest/FirestoreQuestRepository.kt [hunk 1]
```diff
@@ -36,13 +37,15 @@ class FirestoreQuestRepository(private val db: FirebaseFirestore) : QuestReposit
...
       Quest(index = index, title = title, description = description, videoPath = videoPath)
     } catch (e: Exception) {
       null
     }
   }
+
+  internal fun fetchVideo(word: String): String {
+    val sanitizedWord = word.replace(" ", "").lowercase() // Remove spaces and convert to lowercase
+    return "$videoPathBase$sanitizedWord"
+  }
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/model/home/quest/FirestoreQuestRepository.kt:50]: The fetchVideo function lack error handling for cases where the video path is invalid or the video fails to load. 
 
**Suggestion**: Add error listeners and display a fallback UI (for example an error text or a placeholder image) when the video cannot be loaded.

**Comment by ziyadbagh** [comment]
> I tested your branch before diving into the code and noticed a few things that could be improved.
> 
> First, I think it would be better to remove the background sound in the videos to make them more professional.
> 
> Additionally, in the video for the word "Hello," the letter H is missing. I'll keep reviewing to see if I spot other missing letters. However, I suggest that before finalizing the app, we divide the list of words among the team to ensure all the videos are accurate and error-free.
> 
> I've submitted a change request for this, but I see you've already spent a lot of time on this feature. Let me know if you'd prefer to address these video-related improvements in a separate task.

Thank you for the feedback! Regarding the sound in the videos, I hadn’t noticed it since my laptop is always on mute. I agree it needs to be fixed, and once the changes are made, I’ll let you know.

File: app/src/test/java/com/github/se/signify/model/home/quest/FirestoreQuestRepositoryTest.kt [hunk 2]
```diff
@@ -285,4 +285,10 @@ class FirestoreQuestRepositoryTest {
     verify(mockTask).addOnCompleteListener(captor.capture())
     captor.value.onComplete(mockTask)
   }
+
+  @Test
+  fun fetchVideo_constructsCorrectPath() {
```
Comment (nada-ezziti) [inline app/src/test/java/com/github/se/signify/model/home/quest/FirestoreQuestRepositoryTest.kt:290]: According to my previous comment, it's more secure to add tests covering the failure cases. 
So we can test scenarios where the videoPath in a Quest is empty, invalid, or inaccessible.

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
The PR is well-structured and demonstrates a strong implementation of the enhanced quest system. The new features, such as dynamic video integration and the immersive fingerspelling dialog, are excellent additions that enhance user experience. 

I’ve requested a change for the fetchVideo function to add error handling and ensure it properly manages invalid inputs. 

Other than that, great job on this PR—well done!

File: app/src/test/java/com/github/se/signify/model/home/quest/FirestoreQuestRepositoryTest.kt
Comment (nada-ezziti) [inline app/src/test/java/com/github/se/signify/model/home/quest/FirestoreQuestRepositoryTest.kt:1]: Good job on achieving a 100% line coverage for this file.

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
SonarCloud raises 2 errors regarding unused strings. After checking, I noticed that the strings "Let's go" and "Locked" have been declared twice under different names. You can safely remove the two strings flagged by SonarCloud from the `strings.xml` files.

File: app/src/main/res/raw/hello.mp4
Comment (ziyadbagh) [reply app/src/main/res/raw/hello.mp4:1]: Its fixed

**Comment by ziyadbagh** [comment]
> SonarCloud raises 2 errors regarding unused strings. After checking, I noticed that the strings "Let's go" and "Locked" have been declared twice under different names. You can safely remove the two strings flagged by SonarCloud from the `strings.xml` files.

Done.

**Review by nada-ezziti — APPROVED** [review]
You've successfully addressed all the requested changes, and everything looks good now. I approve this PR.

**Pull request merged**